src/pass/bound_checker.cc

```diff
@@ -0,0 +1,135 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file bounds_checker.cc
+ */
+// Instrument checker for out of bounds access.
+
+#include <tvm/ir.h>
+#include <tvm/ir_mutator.h>
+#include <tvm/ir_pass.h>
+#include <tvm/ir_visitor.h>
+#include <vector>
+#include <unordered_map>
+#include <utility>
+#include "bound_checker.h"
+
+namespace tvm {
+namespace ir {
+
+class BoundChecker : public IRMutator {
+ public:
+  BoundChecker() {}
+
+  Stmt Mutate_(const Allocate *op, const Stmt &s) final {
+    if (UpdateIsNeeded(op->buffer_var)) {
+      Update(op->buffer_var, op->extents);
+    }
+    return IRMutator::Mutate_(op, s);
+  }
+
+  Expr Mutate_(const Call *op, const Expr &ex) final {
+    if (proceed_store && op->is_intrinsic(intrinsic::tvm_if_then_else)) {
+      unsafe_rewrited = true;
+    }
+    return IRMutator::Mutate_(op, ex);
+  }
+
+  Stmt Mutate_(const Store *op, const Stmt &s) final {
+    bound_collector.clear();
+    proceed_store = true;
+    unsafe_rewrited = false;
+    IRMutator::Mutate_(op, s);
+    proceed_store = false;
+    // Store should has at least one load.
+    if (CanInstrument(op->index, op->buffer_var) && bound_collector.size()) {
+      Collect(op->index, op->buffer_var);
+      Expr condition = MakeCondition();
+      if (!condition.as<StringImm>()) {
+        Stmt nop = Evaluate::make(1);
+        Stmt then_case =
+            Store::make(op->buffer_var, op->value, op->index, op->predicate);
+        Stmt else_case =
+            AssertStmt::make(condition, StringImm::make(error_message), nop);
+        Stmt body = IfThenElse::make(condition, then_case, else_case);
+        return body;
+      }
+    }
+    return s;
+  }
+
+  Expr Mutate_(const Load *op, const Expr &ex) final {
+    if (CanInstrument(op->index, op->buffer_var)) {
+      Collect(op->index, op->buffer_var);
+    }
+    return IRMutator::Mutate_(op, ex);
+  }
+
+ private:
+  bool UpdateIsNeeded(const VarExpr &buffer_var) const {
+    std::lock_guard<std::mutex> lock(BoundCheckerManager::Global()->mutex);
+    return (
+        buffer_var.defined() &&
+        BoundCheckerManager::Global()->mem_to_buffer.count(buffer_var.get()));
+  }
+
+  void Update(const VarExpr &buffer_var, const Array<Expr> &new_shape) {
+    std::lock_guard<std::mutex> lock(BoundCheckerManager::Global()->mutex);
+    BoundCheckerManager::Global()->mem_to_buffer[buffer_var.get()] = new_shape;
+  }
+
+  bool CanInstrument(const Expr &index, const VarExpr &buffer_var) const {
+    std::lock_guard<std::mutex> lock(BoundCheckerManager::Global()->mutex);
+    return index.defined() && buffer_var.defined() &&
+           BoundCheckerManager::Global()->mem_to_buffer.count(
+               buffer_var.get()) &&
+           // FIXME. Should we implement check when index is ramp ?
```

Comment:
that's a good point, it may be a good idea to add support for vectorized loops

Reply:
Thanks for mention this. I've added support for vectorized loop when index is ramp.